crypto_box.js

```diff
@@ -95,6 +103,159 @@ function crypto_box_seal_open (m, c, pk, sk) {
   return crypto_secretbox_open_easy(m, c.subarray(epk.length), n, k)
 }
 
+// int
+// crypto_box_curve25519xsalsa20poly1305_beforenm(unsigned char *k,
+//                                                const unsigned char *pk,
+//                                                const unsigned char *sk)
+// {
+//     static const unsigned char zero[16] = { 0 };
+//     unsigned char s[32];
+//
+//     if (crypto_scalarmult_curve25519(s, sk, pk) != 0) {
+//         return -1;
+//     }
+//     return crypto_core_hsalsa20(k, zero, s, NULL);
+// }
+function crypto_box_beforenm (k, pk, sk) {
+  const zero = new Uint8Array(16)
+  const s = new Uint8Array(32)
+
+  if (crypto_scalarmult(s, sk, pk) !== 0) {
+    return -1
+  }
+
+  return xsalsa20.core_hsalsa20(k, zero, s, null)
+}
+
+// int
+// crypto_box_detached_afternm(unsigned char *c, unsigned char *mac,
+//                             const unsigned char *m, unsigned long long mlen,
+//                             const unsigned char *n, const unsigned char *k)
+// {
+//     return crypto_secretbox_detached(c, mac, m, mlen, n, k);
+// }
+function crypto_box_detached_afternm (c, mac, m, n, k) {
+  return crypto_secretbox_detached(c, mac, m, n, k)
+}
+
+// crypto_box_detached(unsigned char *c, unsigned char *mac,
+//                     const unsigned char *m, unsigned long long mlen,
+//                     const unsigned char *n, const unsigned char *pk,
+//                     const unsigned char *sk)
+// {
+//     unsigned char k[crypto_box_BEFORENMBYTES];
+//     int           ret;
+//
+//     COMPILER_ASSERT(crypto_box_BEFORENMBYTES >= crypto_secretbox_KEYBYTES);
+//     if (crypto_box_beforenm(k, pk, sk) != 0) {
+//         return -1;
+//     }
+//     ret = crypto_box_detached_afternm(c, mac, m, mlen, n, k);
+//     sodium_memzero(k, sizeof k);
+//
+//     return ret;
+// }
+function crypto_box_detached (c, mac, m, n, pk, sk) {
+  check(mac, crypto_box_MACBYTES)
+  check(n, crypto_box_NONCEBYTES)
+  check(pk, crypto_box_PUBLICKEYBYTES)
+  check(sk, crypto_box_SECRETKEYBYTES)
+
+  const k = new Uint8Array(crypto_box_BEFORENMBYTES)
+
+  if (crypto_box_beforenm(k, pk, sk) !== 0) {
+    return -1
+  }
+
+  return crypto_box_detached_afternm(c, mac, m, n, k)
+}
+
+// int
+// crypto_box_easy(unsigned char *c, const unsigned char *m,
+//                 unsigned long long mlen, const unsigned char *n,
+//                 const unsigned char *pk, const unsigned char *sk)
+// {
+//     if (mlen > crypto_box_MESSAGEBYTES_MAX) {
+//         sodium_misuse();
+//     }
+//     return crypto_box_detached(c + crypto_box_MACBYTES, c, m, mlen, n,
+//                                pk, sk);
+// }
+function crypto_box_easy (c, m, n, pk, sk) {
+  assert(c.length <= crypto_box_MESSAGEBYTES_MAX, "m should not be more than 'crypto_box_MESSAGEBYTESMAX' bytes")
+
+  return crypto_box_detached(c + crypto_box_MACBYTES, c, m, n, pk, sk)
```

Comment:
I'm confused by `c + crypto_box_MACBYTES` here -- I thought that `c` was the buffer/uint8 we mutated to provide output, but it seems that we're trying to add to it? 

```javascript
const c = new Uint8Array(16);
const crypto_box_MESSAGEBYTES_MAX = Number.MAX_SAFE_INTEGER;

c + crypto_box_MESSAGEBYTES_MAX // => "0,0,0,0,0,0,0,0,0,09007199254740991"
```

This is definitely not what we want.

Reply:
Oh, duh -- `c` is a pointer. But from the perspective of `crypto_box_detached` in the original C:

- **`*c`** -- `c + crypto_box_MACBYTES`
- **`*mac`** -- `c`

Assuming `c` starts off empty, the MAC is going to be 16 zeroes, which remain prefixed at `c`. If that's true, maybe I should be passing `c` and trusting that `crypto_secretbox_detached()` is going to format the output for me.

Reply:
Generally, for function calls where some `len` is specified, we have been passing explicit subarrays instead, like so:

```js
crypto_box_detached(c.subarray(crypto_box_MACBYTES), c.subarray(0, crypto_box_MACBYTES), m, n, pk, sk)
```